generic/tclBinary.c

```diff
@@ -183,6 +183,9 @@ typedef struct ByteArray {
 				 * array. */
     unsigned int allocated;	/* The number of bytes allocated for storage
 				 * of the following "bytes" field. */
+#ifdef __CHERI_PURE_CAPABILITY__
+    _Alignas(max_align_t)
```

Comment:
As far as I can tell, there's no obvious way in which a bytearray is expected to be convertible back into an object that contains pointers? Is there a real case you've found of that?

Reply:
Yes, as far as I can recall it leaded to crash caused by tag stripping when running sqlite3 tests. I need to find this test again and reproduce this scenario. Will be back with more information.

Reply:
It looks like sqlite3 stores a pointer in this buffer but I cannot see anything within tcl that would clearly say that it's allowed. The comment above ByteArray suggests ByteArray holds a sequence of one-byte values.

We'd have to decide if we want to fix tcl to be compatible with its users that overuse ByteArray or we'd require all users of tcl to be fixed. If a user wants to store a pointer in ByteArray, they can still make sure they do that in a pointer-aligned location of ByteArray.

Reply:
1. Byte-array objects are typically used to hold the results of binary IO operations or data structures created with the binary command. [...] a byte-array is an array of 8-bit quantities with no implicit meaning. [Tcl8.6.14/Tk8.6.14 Documentation](https://www.tcl-lang.org/man/tcl/TclLib/ByteArrObj.htm)
2. Failing sqlite test:
```tcl
set simple [db one {SELECT fts3_tokenizer('simple')}]
#...
do_test fts3atoken-1.7 {
  execsql {
    SELECT fts3_tokenizer('blah2', $simple) IS NULL;
  }
} {1}
```
3. From `ext/fts3/README.tokenizers`
>  The fts3_tokenizer() function may be called with one or two arguments,
>  as follows:
>
>    SELECT fts3_tokenizer(\<tokenizer-name>);
>    SELECT fts3_tokenizer(\<tokenizer-name>, <sqlite3_tokenizer_module ptr>);
>  
>  Where \<tokenizer-name> is a string identifying the tokenizer and  <sqlite3_tokenizer_module ptr> is a **pointer to an sqlite3_tokenizer_module  structure encoded as an SQL blob**. [...] If only one argument is passed, a pointer to the tokenizer implementation currently registered as <tokenizer-name> is returned, encoded as a blob. [...]
4. During the first call to `fts3_tokenizer()`, the query returns SQL blob with valid pointer, but it is then memcopied to Tcl `ByteArray`, where it is stored at unaligned address [[tclsqlite.c]](https://github.com/CTSRD-CHERI/sqlite/blob/d87461ac449eec2a6165942439da7e033ddaae9f/src/tclsqlite.c) 
```c
    case SQLITE_BLOB: {
      int bytes = sqlite3_column_bytes(pStmt, iCol);
      const char *zBlob = sqlite3_column_blob(pStmt, iCol);
      if( !zBlob ) bytes = 0;
      return Tcl_NewByteArrayObj((u8*)zBlob, bytes);
    }
```
```gdb
(gdb) p *(void**)zBlob
$8 = (void *) 0x3c2320 <simpleTokenizerModule> [rR,0x3c2320-0x3c2390]
(gdb) bt
#0  0x0000000040956dd0 in memcpy (dst0=0x41a8dd68 [rwRW,0x41a8d000-0x41a91000], src0=<optimized out>, length=16)
#1  0x00000000404e3ec0 in Tcl_NewByteArrayObj ()
```
Then it's copied to the variable
```c
data = Tcl_GetByteArrayFromObj(pVar, &n);
(gdb) p *(void**)data
$3 = (void *) 0x3c2320 <simpleTokenizerModule> [rR,0x3c2320-0x3c2390] (invalid)
```

An attempt to use it later fails with Invalid address alignment:
```c
pPtr = *(void **)sqlite3_value_blob(argv[1]); 
```
```gdb
Program received signal SIGBUS, Bus error.
Invalid address alignment.
0x00000000002fda28 in fts3TokenizerFunc (context=0x41a686a0 [rwRW,0x41a5d000-0x41a68b80], argc=2, argv=0x41a686f0 [rwRW,0x41a5d000-0x41a68b80]) at sqlite3.c:190336

> 0x2fda28 <fts3TokenizerFunc+268>        ldr     c0, [c0]
(gdb) p *(void**)$c0
$5 = (void *) 0x3c2320 <simpleTokenizerModule> [rR,0x3c2320-0x3c2390] (invalid)
```

Reply:
Hm, this feels like an abuse of the Tcl API, but it's not clear.

Reply:
I've dropped this commit as we're not sure whether it's a problem with Tcl or sqlite .

Reply:
It would be good to know what Tcl and SQLite developers think about this. Alternatively, if we find multiple cases where it's overused, it might be easier to fix Tcl to allow storing capabilities in ByteArray than fix all users of this data structure.